conservative rounding directions in relations

Shoeboxam opened the pull request:
Closes #150 (main issue) This PR is an initial pass at making the library robust to overflow and rounding in arithmetic.

There were two other issues that were also related to rounding:
Closes #314 
Closes #343

This PR does not add relation relaxation terms for worst-case floating-point rounding in the function (#348).

Shoeboxam commented:
This became pretty nasty, because I needed to define forward maps with specific rounding modes in division. I decided to "do it right" for floats as well. I think I should give this some time to look for a better approach.

andrewvyrros commented on rust/opendp/src/samplers.rs:
```diff
@@ -502,17 +513,21 @@ impl<T: CastInternalReal + SampleRademacher + Zero> SampleLaplace for T {
-        let noised = shift.mul_add(&scale.clone().recip(), &laplace);
-        // postprocessing remains differentially private
-        Ok(Self::from_internal(noised * scale))
+        use rug::float::Round;
+        use rug::ops::{DivAssignRound, AddAssignRound};
+
+        // (shift / scale + noise) * scale.
+        let mut value = shift.into_internal();
+        // when scaling into the noise coordinate space, round down so that noise is overestimated
+        value.div_assign_round(&scale.clone().into_internal(), Round::Zero);
+        // the noise itself is never scaled. Round away from zero to offset the scaling bias
+        value.add_assign_round(
+            &laplace, if value.is_sign_positive() {Round::Up} else {Round::Down});
+        // postprocessing back to original coordinate space
+        //     remains differentially private via postprocessing
+        Ok(Self::from_internal(value) * scale)
```
Should we factor this out, so that gaussian and Laplace can both use same logic?

Shoeboxam replied:
Factored out, added some more comments, and also censored the mpfr flags, FWIW.

andrewvyrros commented on rust/opendp/src/trans/mean/mod.rs:
```diff
@@ -1,34 +1,36 @@
         AbsoluteDistance::default(),
-        StabilityRelation::new_from_constant((upper - lower) / _size / _2)))
+        StabilityRelation::new_from_forward(
+            move |d_in: &IntDistance| T::inf_cast(d_in / 2)
```
Should be `d_in.inf_div(2)`?

Shoeboxam replied:
This one is ok, owing to the newfangled code comments. When odd, we don't consider datasets with an odd number of edits, so we need to compute `(d_in - 1) / 2` anyways.

andrewvyrros commented on rust/opendp/src/trans/sum/mod.rs:
```diff
@@ -22,30 +22,34 @@ pub fn make_bounded_sum<T>(
-        // d_out >= d_in * (M - m) / 2
-        StabilityRelation::new_from_constant((upper - lower) / T::exact_int_cast(2)?)))
+        StabilityRelation::new_from_forward(
+            move |d_in: &IntDistance| T::inf_cast(d_in / 2)
```
Similarly about `inf_div` here?

andrewvyrros commented on rust/opendp/src/trans/variance/mod.rs:
```diff
@@ -1,31 +1,32 @@
-    }
+    let range = upper.inf_sub(&lower)?;
+    // check for potential overflow
+    range.inf_div(&_2)?.inf_mul(&range.inf_div(&_2)?)?;
```
A little explanation here would help.

Shoeboxam replied:
I loosened the check to be equal to the implicit check when computing the constant, but left it duplicated, because the constant calculation would become sensitive to ordering.

andrewvyrros commented on rust/opendp/src/trans/variance/mod.rs:
```diff
@@ -38,34 +39,35 @@ pub fn make_sized_bounded_variance<T>(
-                / (_size + _1)
-                / (_size - _ddof)
-                / _2)))
+            range.inf_mul(&range)?
```
Is this sufficient, such that we don't need the check above on line 29?

andrewvyrros commented on rust/opendp/src/traits.rs:
```diff
@@ -544,3 +504,328 @@ impl FloatBits for f32 {
+        }
+    })+)
+}
+impl_alerting_abs_float!(f32, f64);
```
In discussion, we proposed splitting this up, so that all traits are in a single file for clarity, and impls/macros are in a separate file.

Shoeboxam replied:
I did some significant refactoring here.

andrewvyrros reviewed (commented):
This is great stuff!

I agree with your suggestion that we merge the InfXXX and NegInfXXX variations.

Shoeboxam commented on rust/opendp/src/samplers.rs:
```diff
@@ -398,10 +400,13 @@ impl<T: Clone + SampleGeometric + Sub<Output=T> + Bounded + Zero + One + TotalOr
-        let geometric = T::sample_geometric(shift.clone(), direction,1. - alpha, trials)?;
+        // make prob conservatively larger
+        let geometric = T::sample_geometric(
+            shift.clone(), direction,(1.).inf_sub(&alpha)?, trials)?;
```
I made another logical pass through this and significantly changed everything. It worked out nicely.

Shoeboxam commented:
I did some significant shuffling of the traits, and added a call to censor_flags from mpfr. I also adjusted the tests to use an input distance of 2 when working with sized domains. A surprising side-effect of the new behavior is that on transformations or measurements with an input SizedDomain, `.check(1, 0.)` will always pass. This is because the input dataset distance is divided down to zero. It's technically right, but has the potential to be a footgun. The following will pass:

```python
assert (make_sized_bounded_mean() >> make_base_laplace(0.)).check(1, 0.) 
```

andrewvyrros commented:
> I also adjusted the tests to use an input distance of 2 when working with sized domains. A surprising side-effect of the new behavior is that on transformations or measurements with an input SizedDomain, `.check(1, 0.)` will always pass. This is because the input dataset distance is divided down to zero. It's technically right, but seems like a footgun. The following will pass:
> 
> ```python
> assert (make_sized_bounded_mean() >> make_base_laplace(0.)).check(1, 0.) 
> ```

Yeah this is a funny quirk, maybe could use a note in the docs?

andrewvyrros reviewed (approved):
For such a big, tricky change, this came out nicely! I'm really glad we spent the time to walk through this together, it helped me understand it much better.

Breaking up the traits was a good move, makes things more manageable. Also the cleanup in `samplers.rs` is nice.

It feels like this work puts us in pretty good shape regarding floating point issues.

Shoeboxam commented:
@andrewvyrros It does! The only floating-point issue I'm currently aware of is #348.
